File: Textbook Exercises/Python/Chapter 3/Exercise 3.11/nonconvex_logistic_growth.py
# This file is associated with the book
# "Machine Learning Refined", Cambridge University Press, 2016.
# by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

from mpl_toolkits.mplot3d import Axes3D
from pylab import *

# load the data
def load_data():
    # load data
    global X,y,ax1,ax2

    data = matrix(genfromtxt('bacteria_data.csv', delimiter=','))
    x = asarray(data[:,0])
    temp = ones((size(x),1))
    X = concatenate((temp,x),1)
    y = asarray(data[:,1])
    y = y/y.max()

    # initialize figure, plot data, and dress up panels with axes labels etc.,
    fig = plt.figure(facecolor = 'white')
    ax1 = fig.add_subplot(121)
    ax1.set_xlabel('$x$',fontsize=20,labelpad = 20)
    ax1.set_ylabel('$y$',fontsize=20,rotation = 0,labelpad = 20)
    plot(x,y,'ko')
    ax1.set_xlim(min(x[:,0])-0.5, max(x[:,0])+0.5)
    ax1.set_ylim(min(y)-0.1,max(y)+0.1)

    ax2 = fig.add_subplot(122, projection='3d')
    ax2.xaxis.set_rotate_label(False)
    ax2.yaxis.set_rotate_label(False)
    ax2.zaxis.set_rotate_label(False)
    ax2.get_xaxis().set_ticks([-3,-1,1,3])
    ax2.get_yaxis().set_ticks([-3,-1,1,3])
    ax2.set_xlabel('$w_0$   ',fontsize=20,rotation = 0,linespacing = 10)
    ax2.set_ylabel('$w_1$',fontsize=20,rotation = 0,labelpad = 50)
    ax2.set_zlabel('   $g(\mathbf{w})$',fontsize=20,rotation = 0,labelpad = 20)

###### ML Algorithm functions ######
# run gradient descent
def gradient_descent(w0):
    w_path = []         # container for weights learned at each iteration
    obj_path = []       # container for associated objective values at each iteration
    w_path.append(w0)
    obj = calculate_obj(w0)
    obj_path.append(obj)
    grad_path = []
    w = w0

    # start gradient descent loop
    grad = 1
    iter = 1
    max_its = 5000
    alpha = 10**(-2)
    while linalg.norm(grad) > 10**(-5) and iter <= max_its:
        # compute gradient
# --->  grad =

        # take gradient step
        w = w - alpha*grad

        # update path containers
        w_path.append(w)
        obj = calculate_obj(w)
        obj_path.append(obj)
        iter+= 1

    # reshape containers for use in plotting in 3d
    w_path = asarray(w_path)
    w_path.shape = (iter,2)
    obj_path = asarray(obj_path)
    obj_path.shape = (iter,1)

    return (w_path,obj_path)
    ## for use in testing if gradient vanishes
    # grad_path = asarray(grad_path)
    # grad_path.shape = (iter,1)
    # plot(asarray(grad_path))
    # show()

# calculate the objective value for a given input weight w
def calculate_obj(w):
    temp = 1/(1 + my_exp(-dot(X,w))) - y
    temp = dot(temp.T,temp)
    return temp

# avoid overflow when using exp - just cutoff after arguments get too large/small
def my_exp(u):
    s = argwhere(u > 100)
    t = argwhere(u < -100)
    u[s] = 0
    u[t] = 0
    u = exp(u)
    u[t] = 1
    return u

###### plotting functions #######
# make 3d surface plot
def plot_logistic_surface():
    r = linspace(-3,3,150)
    s,t = meshgrid(r, r)
    s = reshape(s,(size(s),1))
    t = reshape(t,(size(t),1))
    h = concatenate((s,t),1)

    # build 3d surface
    surf = zeros((size(s),1))
    max_its = size(y)
    for i in range(0,max_its):
        surf = surf + add_layer(X[i,:],y[i],h)


    s = reshape(s,(sqrt(size(s)),sqrt(size(s))))
    t = reshape(t,(sqrt(size(t)),sqrt(size(t))))
    surf = reshape(surf,(sqrt(size(surf)),sqrt(size(surf))))

    # plot 3d surface
    ax2.plot_surface(s,t,surf,cmap = 'jet')
    ax2.azim = 175
    ax2.elev = 20

# used by plot_logistic_surface to make objective surface of logistic regression cost function
def add_layer(a,b,c):
    a.shape = (2,1)
    b.shape = (1,1)
    z = my_exp(-dot(c,a))
    z = 1/(1 + z) - b
    z = z**2
    return z

# plot fit to data and corresponding gradient descent path onto the logistic regression objective surface
def show_fit_and_paths(w_path,obj_path,col):
    # plot solution of gradient descent fit to original data
    s = linspace(0,X.max(),100)
    t = 1/(1 + my_exp(-(w_path[-1,0] + w_path[-1,1]*s)))
    ax1.plot(s,t,color = col)

    # plot grad descent path onto surface
    ax2.plot(w_path[:,0],w_path[:,1],obj_path[:,0],color = col,linewidth = 5)   # add a little to output path so its visible on top of the surface plot

def main():

    load_data() # load in data

    plot_logistic_surface()  # plot objective surface

    ### run gradient descent with first initial point
    w0 = array([0,2])
    w0.shape = (2,1)
    w_path, obj_path = gradient_descent(w0)

    # plot fit to data and path on objective surface
    show_fit_and_paths(w_path, obj_path,'m')

    ### run gradient descent with first initial point
    w0 = array([0,-2])
    w0.shape = (2,1)
    w_path, obj_path = gradient_descent(w0)

    # plot fit to data and path on objective surface
    show_fit_and_paths(w_path, obj_path,'c')
    show()

main()

File: mlrefined_exercises/ed_1/MATLAB/Chapter_3/Exercise_3_11/nonconvex_logistic_growth.m
function nonconvex_logistic_growth()

% This file is associated with the book
% "Machine Learning Refined", Cambridge University Press, 2016.
% by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

% load data, plot data and surfaces
[X,y] = load_data();
[s,t,non_obj] = plot_surface(X,y);
plot_pts(X,y);

%%% run grad descent with 2 starting points and plot descent path %%%

% run grad descent with first starting point
w0 = [0;2];
[in,out] = grad_descent(X,y,w0);

% plot results
plot_sigmoid(X,y,in(end,:),1);
plot_descent_path(in,out,1,s,t,non_obj);

% run grad descent with second starting point
w0 = [0;-2];
[in,out] = grad_descent(X,y,w0);

% plot results
plot_sigmoid(X,y,in(:,end),2)
plot_descent_path(in,out,2,s,t,non_obj)

% perform grad descent
function [in,out] = grad_descent(X,y,w)
    % step length
    alpha = 10^-2;

    % Initializations
    in = [w];
    out = [norm(1./(1 + exp(-X*w)) - y)^2];
    grad = 1;
    iter = 1;
    max_its = 10000;

    % main loop
    while  norm(grad) > 10^-5 && iter < max_its

        % fixed steplength
% ----> grad = 
        w = w - alpha*grad;

        % update containers
        in = [in w];
        out = [out ; norm(1./(1 + exp(-X*w)) - y)^2];
        iter = iter + 1;
    end
    in = in';
end

function [s,t,non_obj] = plot_surface(X,y)
    % setup surface
    range = 3;                     % range over which to view surfaces
    [s,t] = meshgrid(-range:0.2:range);
    s = reshape(s,numel(s),1);
    t = reshape(t,numel(t),1);
    non_obj = zeros(length(s),1);   % nonconvex surface

    % build surface
    for i = 1:length(y)
        non_obj = non_obj + non_convex(X(i,:),y(i),s,t)';
    end

    % plot surface
    figure(2)
    subplot(1,2,1)
    set(get(gca,'YLabel'),'Rotation',0)
    set(gcf,'color','w');
    set(gcf,'color','w');
    r = sqrt(numel(s));
    s = reshape(s,r,r);
    t = reshape(t,r,r);
    non_obj = reshape(non_obj,r,r);
    surf(s,t,non_obj)
    box on
    xlabel('w','Fontsize',18,'FontName','cmmi9')
    ylabel('b','Fontsize',18,'FontName','cmmi9')
    zlabel('g','Fontsize',18,'FontName','cmmi9')
    set(get(gca,'ZLabel'),'Rotation',0)

    % plot contour
    subplot(1,2,2)
    set(gcf,'color','w');
    r = sqrt(numel(s));
    s = reshape(s,r,r);
    t = reshape(t,r,r);
    non_obj = reshape(non_obj,r,r);
    contourf(s,t,non_obj,10)
    box on
    xlabel('w','Fontsize',18,'FontName','cmmi9')
    ylabel('b','Fontsize',18,'FontName','cmmi9')
    set(get(gca,'YLabel'),'Rotation',0)
end

function plot_pts(X,y)
    % plot labeled points
    figure(1)
    scatter(X(:,2),y,'fill','k')
    set(gcf,'color','w');
    xlabel('x','Fontsize',18,'FontName','cmmi9')
    ylabel('y','Fontsize',18,'FontName','cmmi9')
    set(get(gca,'YLabel'),'Rotation',0)
    set(gcf,'color','w');
    box on
    set(gca,'FontSize',12);
end

function plot_sigmoid(X,y,w,i)
    % plot
    figure(1)
    hold on
    u = [0:0.1:max(X(:,2))];
    w = 1./(1+exp(-(u*w(2) + w(1))));
    if i == 1
        plot(u,w,'m','LineWidth',2);
    else
        plot(u,w,'g','LineWidth',2);
    end
end

function plot_descent_path(in,out,i,s,t,non_obj)

    % plot nonconvex-output path
    figure(2)
    subplot(1,2,1)
    hold on
    if i == 1
       plot3(in(:,1),in(:,2),out,'m','LineWidth',3);
    else
       plot3(in(:,1),in(:,2),out,'g','LineWidth',3);
    end
    axis([min(min(t)) max(max(t)) min(min(s)) max(max(s)) min(min(non_obj)) max(max(non_obj))])
    subplot(1,2,2)
    hold on
    if i == 1
       plot3(in(:,1),in(:,2),out,'m','LineWidth',3);
    else
       plot3(in(:,1),in(:,2),out,'g','LineWidth',3);
    end
end

% loads data for processing
function [X,y] = load_data()
    % load bacteria data
    data = csvread('bacteria_data.csv');
    x = data(:,1);
    y = data(:,2);
    X = [ones(length(x),1) x];
end

function s = non_convex(c,z,s,t)    % objective function for nonconvex problem
    s = (sigmoid(c*[s,t]') - z).^2;
end

function y = sigmoid(z)
y = 1./(1+exp(-z));
end

end
